libraries/Adapters/Microsoft.Bot.Builder.Adapters.Twilio/TwilioAdapter.cs

```diff
@@ -0,0 +1,243 @@
+                }
+            }
+
+            return responses.ToArray();
+        }
+
+        /// <summary>
+        /// Accept an incoming webhook request and convert it into a TurnContext which can be processed by the bot's logic.
+        /// </summary>
+        /// <param name="request">A request object from Restify or Express.</param>
+        /// <param name="response">A response object from Restify or Express.</param>
+        /// <param name="bot">A bot with logic function in the form `async(context) => { ... }`.</param>
+        /// <param name="cancellationToken">A cancellation token for the task.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public async Task ProcessAsync(HttpRequest request, HttpResponse response, IBot bot, CancellationToken cancellationToken = default)
+        {
+            response.StatusCode = 200;
+            await response.WriteAsync(string.Empty);
+
+            var twilioSignature = request.Headers["x-twilio-signature"];
+
+            var validationUrl = _options.ValidationUrl ?? (request.Headers["x-forwarded-proto"][0] ?? request.Protocol + "://" + request.Host + request.Path);
+
+            var requestValidator = new RequestValidator(_options.AuthToken);
+
+            var bodyStream = new StreamReader(request.Body);
+
+            dynamic payload = bodyStream.ReadToEnd();
+
+            var formattedPayload = payload.Replace("+", "%20");
+
+            var body = QueryStringToDictionary(formattedPayload);
+
+            if (!twilioSignature.Equals(string.Empty) && requestValidator.Validate(validationUrl, body, twilioSignature))
+            {
+                var jsonBody = JsonConvert.SerializeObject(body);
+
+                TwilioEvent twilioEvent = JsonConvert.DeserializeObject<TwilioEvent>(jsonBody);
+
+                var activity = new Activity()
+                {
+                    Id = twilioEvent.MessageSid,
+                    Timestamp = new DateTime(),
+                    ChannelId = "twilio-sms",
+                    Conversation = new ConversationAccount()
+                    {
+                        Id = twilioEvent.From,
+                    },
+                    From = new ChannelAccount()
+                    {
+                        Id = twilioEvent.From,
+                    },
+                    Recipient = new ChannelAccount()
+                    {
+                        Id = twilioEvent.To,
+                    },
+                    Text = (twilioEvent as dynamic).Body,
+                    ChannelData = twilioEvent,
+                    Type = ActivityTypes.Message,
+                };
+
+                // Detect attachments
+                if (Convert.ToInt32(twilioEvent.NumMedia) > 0)
+                {
+                    // specify a different event type for Botkit
+                    (activity.ChannelData as dynamic).botkitEventType = "picture_message";
```

Comment:
This won't work with the BotBuilder SDK? I just created a similar adapter for Twilio WhatsApp, that maps the Twilio attachments to activity attachments.

```typescript
            // Has attachments?
            if (message.NumMedia && parseInt(message.NumMedia) > 0) {

                activity.attachments = [];
                const amount = parseInt(message.NumMedia);

                for (let i = 0; i < amount; i++) {
                    const attachment = {
                        contentType: message['MediaContentType' + i],
                        contentUrl: message['MediaUrl' + i]
                    };
                    activity.attachments.push(attachment);
                }

            }
```

Reply:
Not sure, what do you think @ceciliaavila?

Also, if we need to use custom property for the eventType, let's drop botKit from the property name. 

Let's review this one in more dtails and see if it can be mapped to a native attachment like @iMicknl suggests

Reply:
We will be reviewing @iMicknl approach to apply to our adapter. By the way, the property name no longer has botkit in it.
Thanks!

Reply:
I like @iMicknl's solution.

Reply:
Thanks, we will be working on applying this to our adapter.